fixes #12127 - Foreman should verify x509 sans when authenticating a smart proxy

timogoebel opened the pull request:
This PR adds the feature, that Foreman verifies the subject alternative name attributes of the client cert if they are set in the client certificate. This requires ''SSLOptions +ExportCertData" in Apache. Will add a pr for that soon.
This helps in a ha environment when using the vipname in the san.

Kudos to @martinpfeifer with whom I enjoyed working on that.

Do we need any tests for that? If so, where and what should we test? :-)

timogoebel committed:
fixes #12127 - Foreman should verify x509 subject alternative names when authenticating a smart proxy

timogoebel commented:
PR for puppet code: https://github.com/theforeman/puppet-foreman/pull/365

mmoll commented:
[test]

timogoebel commented:
Refactored the san extraction a bit.

domcleal commented:
[test]

timogoebel commented:
Fix Robocup issue:

`Useless assignment to variable - `id`. Use `_` or `_id` as a variable name to indicate that it won't be used.`

domcleal commented:
[test]

I've sent you an invitation to the GitHub org if you'd like to join - if you wish to accept, and set the membership to public ([here](https://github.com/orgs/theforeman/people)) then your PRs will get auto-tested.

timogoebel commented:
@domcleal : Thanks. I am honored to accept. ;-) Could you just resend the invitation please? Did not receive it just yet.

domcleal commented:
I must have misclicked, hopefully that did it!

dLobatog commented:
@timogoebel Rubocop's now complaining about empty space between the `!` and whatever follows, sorry!

timogoebel commented:
@dLobatog: Should have seen that one coming, sorry! Fixed.

ohadlevy commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -49,11 +50,37 @@ def require_smart_proxy_or_login(features = nil)
+        cert_raw = request.env[Setting[:ssl_client_cert_env]]
+        certificate = OpenSSL::X509::Certificate.new(cert_raw)
+        subject_alt_name_extension = certificate.extensions.find {|e| e.oid == "subjectAltName"}
+        if subject_alt_name_extension
```
does it make sense to extract to smaller methods? i think its easier to follow the logic this way?

timogoebel replied:
Probably yes, I could extract the san and dn extraction in a separate method. What about something like
dn, subject_alternative_names = extract_dn_and_san_from_certificate(certificate) ?

domcleal replied:
Yes, that would work well I think - it could take most of this conditional block out.

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -49,11 +50,37 @@ def require_smart_proxy_or_login(features = nil)
+      # If we have the client certficate in the request environment we can extract the dn and sans from there
+      # and use the dn in the environment if no certificate is set
+      # SAN validation requires "SSLOptions +ExportCertData" in Apache httpd
+      if request.env.has_key?(Setting[:ssl_client_cert_env]) and !request.env[Setting[:ssl_client_cert_env]].empty?
```
Use `&&` instead of `and` in a conditional, and `.present?` may be clearer than `!...empty?`

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -49,11 +50,37 @@ def require_smart_proxy_or_login(features = nil)
+      # SAN validation requires "SSLOptions +ExportCertData" in Apache httpd
+      if request.env.has_key?(Setting[:ssl_client_cert_env]) and !request.env[Setting[:ssl_client_cert_env]].empty?
+        logger.debug "Examining client certificate to extract dn and sans"
+        dn = nil
```
Should the DN still be retrieved, in addition to the sANs?

timogoebel replied:
I think, it's good practice to either use the dNSNames in the san extension or the cn of the certificate and not use both.

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -49,11 +50,37 @@ def require_smart_proxy_or_login(features = nil)
+        dn = request.env[Setting[:ssl_client_dn_env]]
+      end
+
+      if (dn && dn =~ /CN=([^\s\/,]+)/i) or subject_alternative_names
```
`||` instead of `or`

domcleal commented:
Thanks for the PR @timogoebel, sorry for the delay in reviewing.  It would be very useful to get a test for the SAN extraction, which would be easier if moved to its own method.

I'd suggest generating dummy certificate(s) and putting them into the test/ directory, then directly call your new method to check that it's parsing the SANs out correctly.  A functional test like the others for cert-based authentication could then do `.expects(:extract_dn_and_san_from_certificate).with(..).returns([dn, sans])` or similar to avoid needing to test it fully twice.

timogoebel commented:
@domcleal : I extracted certificate handling to a small class and added unit tests for it. What do you think?

I have some trouble with writing functional tests. I guess, they should go to test/functional/smart_proxy_auth_test.rb. However, there are no tests for the certificate stuff, yet and I am  unable to establish a pattern for the tests. Can you point me in the right direction?

timogoebel commented:
Hm.. tests are failing. Let's have a look...

timogoebel commented:
@dLobatog: Somehow this breaks katello tests, any idea why?

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -50,11 +50,29 @@ def require_smart_proxy_or_login(features = nil)
+            request_hosts << $1 if $1
+          else
+            request_hosts << certificate.subject if certificate.subject and !certificate.subject_alternative_names
+            request_hosts += subject_alternative_names if certificate.subject_alternative_names
```
missing `certificate.` here

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -50,11 +50,29 @@ def require_smart_proxy_or_login(features = nil)
+          unless certificate
+            request_hosts << $1 if $1
+          else
+            request_hosts << certificate.subject if certificate.subject and !certificate.subject_alternative_names
```
Use `&&` instead of `and`.

It might be better to rewrite this as a more basic if/else, since this line checks for absence of SANs, the next checks for presence.

<pre>
if certificate.subject_alternative_names
  request_hosts += certificate.subject_alternative_names
elsif certificate.subject
  request_hosts << certificate.subject
end
</pre>


or similar.

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -50,11 +50,29 @@ def require_smart_proxy_or_login(features = nil)
-          request_hosts = [$1]
+          # If the client sent certificate contains a subject or sans, use them for request_hosts, else fall back to the dn set in the request environment
+          request_hosts = []
+          unless certificate
```
Best to swap the two parts of the conditional around, so we have `if certificate` first (positive, then negative).

domcleal commented on app/models/setting/auth.rb:
```diff
@@ -24,6 +24,7 @@ def self.load_defaults
         self.set('ssl_priv_key', N_("SSL Private Key file that Foreman will use to communicate with its proxies"), ssl_priv_key, N_('SSL private key')),
         self.set('ssl_client_dn_env', N_('Environment variable containing the subject DN from a client SSL certificate'), 'SSL_CLIENT_S_DN', N_('SSL client DN env')),
         self.set('ssl_client_verify_env', N_('Environment variable containing the verification status of a client SSL certificate'), 'SSL_CLIENT_VERIFY', N_('SSL client verify env')),
+        self.set('ssl_client_cert_env', N_('Environment variable containing a client\'s SSL certificate'), 'SSL_CLIENT_CERT', N_('SSL client cert env')),
```
Can you use double quotes here?  We use a code parser to extract strings for translation, and it might be safer than escape characters.. it's pretty dumb.

domcleal commented on app/services/certificate_extract.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def subject
+    @subject || @subject = @certificate.subject.to_s[/CN=([^\s\/,]+)/i, 1]
```
`@subject ||=`

domcleal commented:
> @domcleal : I extracted certificate handling to a small class and added unit tests for it. What do you think?

I like that a lot, much clearer and we'll be able to improve and test that class more in the future.  We can perhaps refactor DN parsing into it too one day.

> I have some trouble with writing functional tests. I guess, they should go to test/functional/smart_proxy_auth_test.rb. However, there are no tests for the certificate stuff, yet and I am unable to establish a pattern for the tests. Can you point me in the right direction?

That surprised me too!  I thought we had some.

I've written a couple of tests, one for SANs and one for basic DN parsing to check that's still working after this change.  It found one bug, also fixed.  Please just squash my commit straight into yours, and change/expand as you wish!

https://github.com/timogoebel/foreman/compare/feature_smart_proxy_san...domcleal:12127-cert-san-tests

> Somehow this breaks katello tests, any idea why?

I think Katello has code to copy test/fixtures/ out of Foreman (@jlsherrill added it, IIRC) which isn't liking the new 'certificates' directory.  Perhaps put it in a different, non-Rails fixture directory.

jlsherrill commented:
I'd be happy to update katello to do a recursive copy on that directory if that is where you'd prefer it @domcleal

domcleal commented:
I'm happy for it to be in a different directory, since that one is a little special by Rails standards.  We have a similar situation in puppet-foreman, so created a separate directory: https://github.com/theforeman/puppet-foreman/tree/master/spec/static_fixtures

timogoebel commented:
@domcleal : Huge thanks for your help, this proves that I am definately more an ops guy than a developer... ;-) Will fix that tomorrow.

domcleal commented:
No worries, you've done a good job!  Merged as 3ba1fa7, thanks @timogoebel.
